### PR TITLE
Upgrade graphql-java version to 13.0

### PR DESCRIPTION
**Pull request by negesti**
**Changes**
* upgrade graphql-java version to 13.0
* replace custom NullValue with graphql.language.NullValue (DataFetcher)

**Deprecated in 13.0**
* Replace new GraphQLScalarType with newScalar Builder
* use Field.transform to change arguments (all field properties are immutable)
* replace DataFetchingEnvironmentBuilder with DataFetchingEnvironmentImpl (graphql internal class!)
* variables have been moved from DataFetchingEnvironment.Environment directly to DataFetchingEnvironment.variables

Is related to #202

### CODE REVIEW

**Comment by negesti** [comment]
With 13.0 the DataFetcher "configuration" have been moved from die Object definition to a separate CodeRegistry [GraphQL Docs](https://www.graphql-java.com/documentation/v12/data-fetching/)

This would provide the possibility to add some kind of post processing to the Auto-Schema generator and replace/override some of the DataFetchers if needed.

 Imho this would be really nice and makes it possible to add additional field and DataFetchers that are not known to the JPA entity

**Comment by igdianov** [comment]
@negesti Can you update `.travis.yml` to replace oraclejdk8 with openjdk8 in your PR because it is failing in Travis job? Thank you!

File: graphql-jpa-query-schema/src/main/java/com/introproventures/graphql/jpa/query/schema/impl/QraphQLJpaBaseDataFetcher.java [hunk 1]
```diff
@@ -900,7 +893,7 @@ protected final DataFetchingEnvironment wherePredicateEnvironment(DataFetchingEn
     @SuppressWarnings( { "unchecked", "rawtypes" } )
     protected Object convertValue(DataFetchingEnvironment environment, Argument argument, Value value) {
         if (value instanceof NullValue) {
-            return null;
+            return value;
```
Comment (igdianov) [inline graphql-jpa-query-schema/src/main/java/com/introproventures/graphql/jpa/query/schema/impl/QraphQLJpaBaseDataFetcher.java]: @negesti What is the reason to change `return null` to `return value`? I also realized I do not have test coverage for it. This should convert GraphQL AST `NullValue` to Java `null`.

File: graphql-jpa-query-schema/src/main/java/com/introproventures/graphql/jpa/query/schema/impl/QraphQLJpaBaseDataFetcher.java [hunk 2]
```diff
@@ -132,15 +126,15 @@ public QraphQLJpaBaseDataFetcher(EntityManager entityManager,
...
+        return getQuery(environment, environment.getField(), true).getResultList();
     }
 
     protected TypedQuery<?> getQuery(DataFetchingEnvironment environment, Field field, boolean isDistinct) {
         CriteriaBuilder cb = entityManager.getCriteriaBuilder();
         CriteriaQuery<?> query = cb.createQuery((Class<?>) entityType.getJavaType());
         Root<?> from = query.from(entityType);
 
-        DataFetchingEnvironment queryEnvironment = DataFetchingEnvironmentBuilder.newDataFetchingEnvironment(environment)
+        DataFetchingEnvironment queryEnvironment = DataFetchingEnvironmentImpl.newDataFetchingEnvironment(environment)
```
Comment (igdianov) [inline graphql-jpa-query-schema/src/main/java/com/introproventures/graphql/jpa/query/schema/impl/QraphQLJpaBaseDataFetcher.java]: @negesti I think it is better to extract `DataFetchingEnvironmentImpl.newDataFetchingEnvironment` into protected class method `newDataFetchingEnvironment`.

File: graphql-jpa-query-schema/src/main/java/com/introproventures/graphql/jpa/query/schema/JavaScalars.java [hunk 3]
```diff
@@ -107,29 +109,37 @@
...
+
     protected static GraphQLScalarType computeGraphQLScalarType(Class<?> key) {
         String typeName = key.getSimpleName();
         String description = typeName+" Scalar Object Type";
         
-        return new GraphQLScalarType(typeName, description, new GraphQLObjectCoercing());
+        return toScalarType(typeName, description ,new GraphQLObjectCoercing());
+    }
+
+    public static <T extends Coercing> GraphQLScalarType toScalarType(String name, String description, T coercing) {
```
Comment (igdianov) [inline graphql-jpa-query-schema/src/main/java/com/introproventures/graphql/jpa/query/schema/JavaScalars.java]: I think it is better to not to expose this method to public. The less api surface the better.

File: graphql-jpa-query-boot-starter/src/test/java/com/introproventures/graphql/jpa/query/starter/GraphQLJpaQueryStarterIT.java [hunk 4]
```diff
@@ -86,11 +86,31 @@ public void testGraphqlArguments() throws JsonParseException, JsonMappingExcepti
...
+  public boolean isDataPresent() {
+    return data != null;
+  }
+
+  @Override
+  public Map<Object, Object> getExtensions() {
+    return extensions;
+  }
+
+  @Override
```
Comment (igdianov) [inline graphql-jpa-query-boot-starter/src/test/java/com/introproventures/graphql/jpa/query/starter/GraphQLJpaQueryStarterIT.java]: @negesti Can you adjust indentation here?

**Comment by negesti** [comment]
https://github.com/negesti/graphql-jpa-query/blob/6b22a21132c3c7b5159e658b82d66029bb6781fd/graphql-jpa-query-schema/src/test/java/com/introproventures/graphql/jpa/query/converter/GraphQLJpaConverterTests.java#L508-L530

The NullValue is covered here. After replacing your custom NullValue with the graphql-java NullValue the "value of null" is no longer null, it is graphql.language.NullValue

**Comment by molexx** [comment]
@igdianov thanks, personally I think I should be ok (other libs I use have been updated to graphql-java v13 too) but I agree it would be safer to bump to 0.4.x. We can see how everyone gets on and update 0.3.x if needed.


Thanks @negesti  but this isn't working for me. To confirm I've pulled and built `/negesti/graphql-jpa-query` `master` (as of `6b22a211`) and whilst graphql-jpa-query-example-simple looks to start up ok, graphiql fails to load the schema and logs the trace below. I've tried openjdk 8 and 11. I see spring-boot is still 2.1.7 but my project is on 2.1.9 and has the same error.

```
2019-10-15 18:00:46.056 ERROR 15410 --- [nio-8080-exec-4] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is graphql.AssertException: Object required to be not null] with root cause

graphql.AssertException: Object required to be not null
        at graphql.Assert.assertNotNull(Assert.java:29) ~[graphql-java-13.0.jar:na]
        at graphql.execution.instrumentation.DocumentAndVariables.<init>(DocumentAndVariables.java:19) ~[graphql-java-13.0.jar:na]
        at graphql.execution.instrumentation.DocumentAndVariables.<init>(DocumentAndVariables.java:13) ~[graphql-java-13.0.jar:na]
        at graphql.execution.instrumentation.DocumentAndVariables$Builder.build(DocumentAndVariables.java:55) ~[graphql-java-13.0.jar:na]
        at graphql.GraphQL.parse(GraphQL.java:578) ~[graphql-java-13.0.jar:na]
        at graphql.GraphQL.parseAndValidate(GraphQL.java:547) ~[graphql-java-13.0.jar:na]
        at graphql.GraphQL.lambda$parseValidateAndExecute$2(GraphQL.java:531) ~[graphql-java-13.0.jar:na]
        at graphql.execution.preparsed.NoOpPreparsedDocumentProvider.getDocument(NoOpPreparsedDocumentProvider.java:13) ~[graphql-java-13.0.jar:na]
        at graphql.GraphQL.parseValidateAndExecute(GraphQL.java:533) ~[graphql-java-13.0.jar:na]
        at graphql.GraphQL.executeAsync(GraphQL.java:502) ~[graphql-java-13.0.jar:na]
        at graphql.GraphQL.execute(GraphQL.java:432) ~[graphql-java-13.0.jar:na]
        at com.introproventures.graphql.jpa.query.schema.impl.GraphQLJpaExecutor.execute(GraphQLJpaExecutor.java:72) ~[graphql-jpa-query-schema-0.3.34-SNAPSHOT.jar:0.3.34-SNAPSHOT]
```

**Comment by igdianov** [comment]
> https://github.com/introproventures/graphql-jpa-query/blob/6b22a21132c3c7b5159e658b82d66029bb6781fd/graphql-jpa-query-schema/src/test/java/com/introproventures/graphql/jpa/query/converter/GraphQLJpaConverterTests.java#L508-L530
> 
> The NullValue is covered here. After replacing your custom NullValue with the graphql-java NullValue the "value of null" is no longer null, it is graphql.language.NullValue

I don't think so, this test case applies to different scenario. Can you try reverting the change. I will look into the coverage for convertValue later.

**Comment by anotender** [comment]
> @molexx @miklemv @anotender @MurrayFurtado Do you see any problems with upgrading to Graphql-java v13? I think I should start and release MINOR 0.4.x version on master branch instead of new PATCH as this will add functionality in a backwards compatible manner. Do you want to support Graphql-java 11 on branch 0.3.x?

@igdianov there is no problem for me with upgrading to the newer version. Given all of the tests are green I am fine with it.

**Comment by molexx** [comment]
Is graphiql working in graphql-jpa-query-example-simple for everyone else? The tests pass and server starts but schema is not returned - stacktrace above.

**Comment by negesti** [comment]
@molexx the problem is fixed with the last commit.

**Comment by negesti** [comment]
@igdianov finally i  get found the problem with NullValue.

https://github.com/introproventures/graphql-jpa-query/blob/5d99c3b4629521ebee2788c6b877250c279c8bf2/graphql-jpa-query-schema/src/main/java/com/introproventures/graphql/jpa/query/schema/impl/QraphQLJpaBaseDataFetcher.java#L900-L905

The value parameter is a `graphql.language.NullValue` and the if checks against the inner class `QraphQLJpaBaseDataFetcher.NullValue` 
The whole if-else block is skipped and end of the method `return value` returns the graphql.language.NullValue.

**Comment by igdianov** [comment]
@negesti  Let's keep the scope of this PR as small as possible to address only the problem at hand, i.e. upgrade to graphql-java 13.  I have created a fork of the original commit in your PR and added a few changes on top of it with feedback from @molexx  and mine: https://github.com/introproventures/graphql-jpa-query/pull/209

After that, we can discuss any use cases and changes to be able to use code registry, if needed.

**Comment by negesti** [comment]
@igdianov First, sorry for all the unneccesary changes of imports and indention. I really have to remove auto-importer and auto-reformat-code..

From my pov the only real difference between 205 and 209 is, that i have replaced all .dataFetcher calls with the GraphQLCodeRegistry and added the codeRegistry to GraphQLSchemaFactoryBean. 

If you don't want to add this inside the pr, we should at least add a comment in the factory bean and remove the code registry from ``GraphQLSchemaAutoConfigurationTest`` as it adds a code registry that is never used

**Comment by igdianov** [comment]
@negesti No worries. I should have explained that I don't want to make any radical changes to schema builder internals in order to be able to support 0.3.x and 0.4.x core branches at the same time. i.e to be able to backport fixes and features if needed. Refactoring internals to use new code registry will make it more difficult.  

Also, let's agree not to refactor existing tests. It makes it more difficult to review PRs. 

I had taken time to review the upgrade to graphql-java v13 and I am happy to merge it without extra refactoring.

**Comment by negesti** [comment]
sure, fully understand that. I have split the test because the problem with the NullValue object took my around 2 hours to fix and i don't wanted to run all the test's all the time :)

The ``GraphQLCodeRegistry`` should be reverted to the master version. At the moment, it add the codeRegistry to the schema but the SchemaFactoryBean doesn't load it. This makes the test even more confusing

**Comment by igdianov** [comment]
Closed via https://github.com/introproventures/graphql-jpa-query/pull/209

**Pull request closed without merging**